### PR TITLE
Port to ROS 2

### LINKED ISSUE #20: ros2-devel  Parameter not declared exception
when I run this on dashing, passing my urdf file to the node, I get an exception:

rclpy.exceptions.ParameterNotDeclaredException: ('Invalid access to undeclared parameter(s)', 'dependent_joints')

I modified JointStatePublisher:get_param to catch that exception and return None, and the node works now.

### PR DESCRIPTION
**Pull request by clalancette**
This PR is a port of joint_state_publisher to ROS 2.  I did the initial port, but got a bunch of help from people along the way (thanks @gonzodepedro, @schornakj, @emersonknapp, @sloretz, and @pbeeson).

Besides the port to ROS 2, this PR does a number of other things to improve the code:

1.  It adds a README describing the subscribed topics, published topics, and parameters.
1.  If the GUI is in use, and the GUI is closed, this now quits the program (which I think is a little more intuitive).
1.  Allows the joint_state_publisher to get its URDF either from the command-line, or from the `/robot_description` topic.  In the latter case, this node will actually do nothing until something is received on `/robot_description`, at which point it will start publishing.  If both the command-line option is specified and the `/robot_description` topic is available, the command-line option takes precedence.

I've done quite some testing on this to make sure all of the options work.  In my testing with a simple URDF, it does seem to work fairly well.  There are a few TODOs left:

1.  ~~Despite the fact that we `allow_undeclared_parameters`, there doesn't seem to be a good way to get at undeclared parameters in the rclpy API.  Thus, we can't really support `dependent_joints` and `zeros` until we figure that out.  That will be a PR for another time (maybe with some core rclpy changes as well).~~  I actually found a way to do this, so this should work.
1.  ~~While `rclpy.Rate` is available in Eloquent, I wanted to make sure this was backwards-compatible with Dashing as well.  So I left the TODO about switching to Rate in there.~~ This no longer applies since we switched to a timer.

This also fixes #20

### CODE REVIEW

**Comment by sloretz** [comment]
> Despite the fact that we allow_undeclared_parameters, there doesn't seem to be a good way to get at undeclared parameters in the rclpy API. Thus, we can't really support dependent_joints and zeros until we figure that out. That will be a PR for another time (maybe with some core rclpy changes as well).

It looks like this can be made to work with `automatically_declare_parameters_from_overrides`. I'm having trouble understanding the meaning of the options to the parameters API, but it seems `automatically_declare_parameters_from_overrides` declares any parameter that is passed via yaml (Not just parameters via `parameter_overrides`), and `allow_undeclared_parameters` doesn't seem to be necessary.

```yaml
/**:
  ros__parameters:
    dependent_joints:
      joint_D: {parent: joint_A, factor: 3 }
      joint_E: {parent: joint_B }
      joint_F: {parent: joint_C, factor: -1 }
```

```python3
import rclpy
from rclpy.node import Node


if __name__ == '__main__':
    rclpy.init()
    node = Node('mock_jsp', automatically_declare_parameters_from_overrides=True)
    print(node.get_parameters_by_prefix('dependent_joints'))
    rclpy.shutdown()
```

```console
$ python3 mockjsp.py --ros-args --params-file ./jsp_params.yaml --
{'joint_D.parent': <rclpy.parameter.Parameter object at 0x7fd4103f9780>, 'joint_D.factor': <rclpy.parameter.Parameter object at 0x7fd4103f9be0>, 'joint_E.parent': <rclpy.parameter.Parameter object at 0x7fd4103a1390>, 'joint_F.parent': <rclpy.parameter.Parameter object at 0x7fd4156cdc88>, 'joint_F.factor': <rclpy.parameter.Parameter object at 0x7fd407c74160>}
```

**Comment by clalancette** [comment]
@sloretz You are totally right, that works.  Thanks!  I've added back in support for dependent joints and zeros in 475e12f1d0f210fa28610434f9f3a9dbb5024ef0

**Comment by clalancette** [comment]
FYI that this is probably on hold until we decide what to do about #31.  Once we've merged that, I'll likely rebase/rewrite this on top of that.

**Comment by clalancette** [comment]
I've now redone this on top of #31 .  Still needs to wait until that is merged, but once that is in we can easily rebase and review this.

**Comment by clalancette** [comment]
All right, the ROS 1 stuff has been merged into noetic-devel.  I've updated the `ros2` branch to be the same as noetic-devel, and rebased this branch on top of that.  @sloretz This one should be ready for a review.

File: joint_state_publisher/README.md
Comment (sloretz) [inline joint_state_publisher/README.md]: While `robot_state_publisher` can `update TF2 transforms for the robot on the fly`, I'd describe `joint_state_publisher`'s contribution as making sure all transforms exist even when a joint doesn't have encoder data.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (sloretz) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: I'd inherit from `rclpy.node.Node` so we could compose `joint_state_publisher` if we ever get around to making a component container for `rclpy`.

File: joint_state_publisher/README.md
Comment (sloretz) [inline joint_state_publisher/README.md]: `<value>` disappears when rendered by github

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (sloretz) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: this feels more like `debug` than `info` to me

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (sloretz) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: nit space after comma `name, param`. Rock-paper-scissors to decide who makes a follow up PR adding a flake8 test?

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (sloretz) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: Recommend adding `allowed values are ...`

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (sloretz) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: I'm more used to seeing `dj[newkey] = {newvalue: param.value}` for the initial value

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (sloretz) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: [Rate is available](https://github.com/ros2/rclpy/blob/3b323b564f307cd84d9cb2640c3de8b42ed9295f/rclpy/rclpy/timer.py#L89), but I would definitely recommend a timer callback instead of rate + looping + spin. That would be required to enable this node to be composed.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (sloretz) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: This only executes a single callback. If multiple things are ready, then acting on that data will be delayed by the sleep below. I strongly recommend using a timer instead of looping.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (sloretz) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: AFAIK, the `console_scripts` entry point doesn't call the function with any arguments, so `input_args` can be deleted.

https://python-packaging.readthedocs.io/en/latest/command-line-scripts.html#the-console-scripts-entry-point

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (sloretz) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: nit depth of 1 seems fine. If multiple `robot_description` updates come in a row, I don't think anything is lost as long as the very last one gets acted on.

File: joint_state_publisher/README.md
Comment (clalancette) [reply joint_state_publisher/README.md]: Yeah, I was trying to get that across with the previous sentence.  But maybe I muddled the waters too much by talking about `robot_state_publisher` here.  I think I'll just remove this sentence, and slightly change the wording in the previous sentence.

Comment (clalancette) [reply joint_state_publisher/README.md]: Fixed in 2917803

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (clalancette) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: Good point.  Fixed in 2917803.

File: joint_state_publisher/README.md
Comment (clalancette) [reply joint_state_publisher/README.md]: Arg.  Removed the problematic brackets in 2917803.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (clalancette) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: Sure.  Fixed in 2917803.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (clalancette) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: Haha.  I avoided it here, but yeah, we can do a follow-up to add flake8.  Anyway, style fixed in 2917803.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (clalancette) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: Fixed in 2917803.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (clalancette) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: Yeah, excellent call.  Fixed in 2917803.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (clalancette) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: Fixed in 2917803.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (clalancette) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: All right.  The input_args thing is a pattern I've seen elsewhere in the codebase, but I don't feel strongly about it.  Fixed in 2917803.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (clalancette) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: Fixed in 2917803.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (clalancette) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: Fixed in d2be54f

**Comment by clalancette** [comment]
@sloretz Thanks for the great review!  Things are now updated as you recommend in almost all cases.  Please take another look when you get a chance.

File: joint_state_publisher/README.md
Comment (sloretz) [inline joint_state_publisher/README.md]: nit, in ROS 2 the CLI tools would say `sensor_msgs/msg/JointState`

File: joint_state_publisher/README.md
Comment (sloretz) [inline joint_state_publisher/README.md]: `sensor_msgs/msg/JointState`

File: joint_state_publisher/README.md
Comment (sloretz) [inline joint_state_publisher/README.md]: `std_msgs/msg/String`

File: joint_state_publisher/README.md
Comment (sloretz) [inline joint_state_publisher/README.md]: `sensor_msgs/msg/JointState`

File: joint_state_publisher/README.md
Comment (sloretz) [inline joint_state_publisher/README.md]: `sensor_msgs/msg/JointState`

File: joint_state_publisher/README.md
Comment (sloretz) [inline joint_state_publisher/README.md:29]: mimiced -> mimicked

File: joint_state_publisher/test/test_copyright.py
Comment (sloretz) [inline joint_state_publisher/test/test_copyright.py]: nit, 2020

File: joint_state_publisher/test/test_xmllint.py
Comment (sloretz) [inline joint_state_publisher/test/test_xmllint.py]: nit 2020

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (sloretz) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: No need to set `sys.argv`, [rclpy does that already](https://github.com/ros2/rclpy/blob/3b323b564f307cd84d9cb2640c3de8b42ed9295f/rclpy/rclpy/__init__.py#L73)

File: joint_state_publisher_gui/joint_state_publisher_gui/joint_state_publisher_gui.py
Comment (sloretz) [inline joint_state_publisher_gui/joint_state_publisher_gui/joint_state_publisher_gui.py]: No need for input_args if the `console_scripts` entry point doesn't pass any arguments.

File: joint_state_publisher_gui/joint_state_publisher_gui/joint_state_publisher_gui.py
Comment (sloretz) [inline joint_state_publisher_gui/joint_state_publisher_gui/joint_state_publisher_gui.py:262]: I launched the GUI with no urdf via CLI, and then published it via `/robot_description`, but the sliders don't show up. Is something supposed to update the GUI when that happens?

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (sloretz) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py:374]: I get an `IndexError` here when I publish the same URDF twice on the `/robot_description` topic. It looks like `self.joint_list` has duplicate joint names, but `num_joints` was calculated from `free_joints` and `dependent_joints` which can't have duplicates because they're dictionaries, so `msg.position` is half the length. Maybe the variables that are set in `__init__` need to be set in `self.configure_robot` instead so they get cleared when a new description is published?

File: joint_state_publisher/README.md
Comment (clalancette) [reply joint_state_publisher/README.md]: Fixed in 2832411.

File: joint_state_publisher/README.md
Comment (clalancette) [reply joint_state_publisher/README.md]: Fixed in Fixed in 2832411.

File: joint_state_publisher/README.md
Comment (clalancette) [reply joint_state_publisher/README.md]: Fixed in 2832411.

File: joint_state_publisher/README.md
Comment (clalancette) [reply joint_state_publisher/README.md]: Fixed in 2832411.

File: joint_state_publisher/README.md
Comment (clalancette) [reply joint_state_publisher/README.md]: Fixed in 2832411.

File: joint_state_publisher/README.md
Comment (clalancette) [reply joint_state_publisher/README.md:29]: Fixed in 2832411.

File: joint_state_publisher/test/test_copyright.py
Comment (clalancette) [reply joint_state_publisher/test/test_copyright.py]: Fixed in 2832411.

File: joint_state_publisher/test/test_xmllint.py
Comment (clalancette) [reply joint_state_publisher/test/test_xmllint.py]: Fixed in 2832411.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (clalancette) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: Fixed in 2832411.

File: joint_state_publisher_gui/joint_state_publisher_gui/joint_state_publisher_gui.py
Comment (clalancette) [reply joint_state_publisher_gui/joint_state_publisher_gui/joint_state_publisher_gui.py]: Fixed in 2832411.

File: joint_state_publisher_gui/joint_state_publisher_gui/joint_state_publisher_gui.py
Comment (clalancette) [reply joint_state_publisher_gui/joint_state_publisher_gui/joint_state_publisher_gui.py:262]: Yep, this was another missing feature.  Thanks for catching.  Fixed in 2832411.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py
Comment (clalancette) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py:374]: This was absolutely a problem.  Fixing it turned out to be...complicated.  Commit 735f2a7 successfully fixes it, along with making the layout way more dynamic and useful (in my opinion).  It also simplifies the code a bit, at the expense of adding a new PyQt layout type (FlowLayout), and putting the Center/Randomize buttons at the top.  I think that is a pretty good tradeoff for a dynamic, reflowable UI, but please take a look and let me know what you think.

**Comment by clalancette** [comment]
@sloretz All of your above points should be fixed now.  Ready for another round, thanks!

**Review by sloretz — APPROVED** [review]

**Pull request merged**
